### PR TITLE
add verify and generate token emulator support

### PR DESCRIPTION
**Pull request by brianrodri**
# `tox` results
**NOTE**: Nondeterministic stuff has been normalized:
```bash
tox |
  sed -E "s/0x[[:xdigit:]]+/0xDEADBEEF/" |
  sed -E "s/PYTHONHASHSEED='[[:digit:]]+'/PYTHONHASHSEED='0'/g" |
  sed -E "s/[[:digit:]]+\.[[:digit:]]+(s| seconds)/00.00\1/g" |
  sed -E "s/^(message = b?)'.*'/\1'XXX'/g"
```

-   `v3.2.1-PATCH` branch results: http://bit.ly/2Mk22gv
-   `py2-emulator-support` branch results: http://bit.ly/3oy3idg
-   ```diff
    --- .tox/base-tox-report.txt	2021-01-30 15:06:16.000000000 -0500
    +++ .tox/head-tox-report.txt	2021-01-30 15:05:03.000000000 -0500
    @@ -126,7 +126,7 @@
     _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
     firebase_admin/auth.py:149: in create_custom_token
         return token_generator.create_custom_token(uid, developer_claims)
    -firebase_admin/_token_gen.py:169: in create_custom_token
    +firebase_admin/_token_gen.py:190: in create_custom_token
         return jwt.encode(signing_provider.signer, payload)
     .tox/py2/lib/python2.7/site-packages/google/auth/jwt.py:112: in encode
         signature = signer.sign(signing_input)
    @@ -166,7 +166,7 @@
     _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
     firebase_admin/auth.py:149: in create_custom_token
         return token_generator.create_custom_token(uid, developer_claims)
    -firebase_admin/_token_gen.py:169: in create_custom_token
    +firebase_admin/_token_gen.py:190: in create_custom_token
         return jwt.encode(signing_provider.signer, payload)
     .tox/py2/lib/python2.7/site-packages/google/auth/jwt.py:112: in encode
         signature = signer.sign(signing_input)
    @@ -317,7 +317,7 @@
     _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
     firebase_admin/auth.py:149: in create_custom_token
         return token_generator.create_custom_token(uid, developer_claims)
    -firebase_admin/_token_gen.py:169: in create_custom_token
    +firebase_admin/_token_gen.py:190: in create_custom_token
         return jwt.encode(signing_provider.signer, payload)
     .tox/py3/lib/python3.9/site-packages/google/auth/jwt.py:112: in encode
         signature = signer.sign(signing_input)
    @@ -357,7 +357,7 @@
     _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
     firebase_admin/auth.py:149: in create_custom_token
         return token_generator.create_custom_token(uid, developer_claims)
    -firebase_admin/_token_gen.py:169: in create_custom_token
    +firebase_admin/_token_gen.py:190: in create_custom_token
         return jwt.encode(signing_provider.signer, payload)
     .tox/py3/lib/python3.9/site-packages/google/auth/jwt.py:112: in encode
         signature = signer.sign(signing_input)
    @@ -414,17 +414,13 @@
       /Users/brianrodri/Repositories/firebase-admin-python/.tox/py3/lib/python3.9/site-packages/google/rpc/status_pb2.py:43: DeprecationWarning: Call to deprecated create function Descriptor(). Note: Create unlinked descriptors is going to go away. Please use get/find descriptors from generated code or query the descriptor_pool.
         _STATUS = _descriptor.Descriptor(
     
    -firebase_admin/_token_gen.py:290
    -  /Users/brianrodri/Repositories/firebase-admin-python/firebase_admin/_token_gen.py:290: SyntaxWarning: "is" with a literal. Did you mean "=="?
    -    if header.get('alg') == 'HS256' and payload.get(
    -
     -- Docs: https://docs.pytest.org/en/stable/warnings.html
     =========================== short test summary info ============================
     FAILED tests/test_db.py::TestDatabaseInitialization::test_valid_db_url[https://test.firebaseio.com]
     FAILED tests/test_db.py::TestDatabaseInitialization::test_valid_db_url[https://test.firebaseio.com/]
     FAILED tests/test_token_gen.py::TestCreateCustomToken::test_sign_with_iam - K...
     FAILED tests/test_token_gen.py::TestCreateCustomToken::test_sign_with_discovered_service_account
    -================= 4 failed, 2366 passed, 11 warnings in 00.00s =================
    +================= 4 failed, 2366 passed, 10 warnings in 00.00s =================
     ERROR: InvocationError for command /Users/brianrodri/Repositories/firebase-admin-python/.tox/py3/bin/pytest (exited with code 1)
     pypy create: /Users/brianrodri/Repositories/firebase-admin-python/.tox/pypy
     pypy installdeps: pytest, pytest-localserver
    @@ -612,7 +608,7 @@
     _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
     firebase_admin/auth.py:149: in create_custom_token
         return token_generator.create_custom_token(uid, developer_claims)
    -firebase_admin/_token_gen.py:169: in create_custom_token
    +firebase_admin/_token_gen.py:190: in create_custom_token
         return jwt.encode(signing_provider.signer, payload)
     .tox/cover/lib/python2.7/site-packages/google/auth/jwt.py:112: in encode
         signature = signer.sign(signing_input)
    @@ -652,7 +648,7 @@
     _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
     firebase_admin/auth.py:149: in create_custom_token
         return token_generator.create_custom_token(uid, developer_claims)
    -firebase_admin/_token_gen.py:169: in create_custom_token
    +firebase_admin/_token_gen.py:190: in create_custom_token
         return jwt.encode(signing_provider.signer, payload)
     .tox/cover/lib/python2.7/site-packages/google/auth/jwt.py:112: in encode
         signature = signer.sign(signing_input)
    @@ -680,12 +676,12 @@
     ----------------------------------------------------------
     firebase_admin/__about__.py                6      0   100%
     firebase_admin/__init__.py               123      2    98%
    -firebase_admin/_auth_utils.py            184      5    97%
    +firebase_admin/_auth_utils.py            207      9    96%
     firebase_admin/_http_client.py            49      1    98%
     firebase_admin/_messaging_encoder.py     357      7    98%
     firebase_admin/_messaging_utils.py       138      0   100%
     firebase_admin/_sseclient.py             116     16    86%
    -firebase_admin/_token_gen.py             203      5    98%
    +firebase_admin/_token_gen.py             240     21    91%
     firebase_admin/_user_import.py           170      0   100%
     firebase_admin/_user_mgt.py              310     11    96%
     firebase_admin/_utils.py                 101      0   100%
    @@ -714,7 +710,7 @@
     tests/test_user_mgt.py                   813      0   100%
     tests/testutils.py                        89      4    96%
     ----------------------------------------------------------
    -TOTAL                                   7558    107    99%
    +TOTAL                                   7618    127    98%
     
     ============== 4 failed, 2366 passed, 1 warnings in 00.00 seconds ==============
     ERROR: InvocationError for command /Users/brianrodri/Repositories/firebase-admin-python/.tox/cover/bin/pytest --cov=firebase_admin --cov=tests (exited with code 1)
    ```

# `./lint.sh all` results
-   ```bash
    Running linter on module firebase_admin
    Running linter on module tests
    Running linter on module integration
    Running linter on module snippets
    ```

### CODE REVIEW

File: firebase_admin/_auth_utils.py [hunk 1]
```diff
@@ -323,3 +340,72 @@ def _build_error_message(code, exc_type, custom_message):
...
+        """Returns the resource URL corresponding to the provided parameters.
+
+        Args:
+            api: The backend API name (optional).
+            **params: The optional additional parameters to substitute in the URL path.
+
+        Returns:
+            string: The corresponding resource URL.
+        """
+        return super(TenantAwareAuthResourceUrlBuilder, self).get_url(
```
Comment (hiranya911) [inline firebase_admin/_auth_utils.py]: You can probably just substitute the tenant_id in the constructor. Then you don't need to override this method.

File: firebase_admin/_token_gen.py [hunk 2]
```diff
@@ -254,98 +275,134 @@ def __init__(self, **kwargs):
...
+                'GOOGLE_CLOUD_PROJECT environment variable.'.format(self.operation))
+
+        header, payload = self._decode_unverified(token)
+
+        self._verify_aud(payload)
+        self._verify_iss(payload)
+        self._verify_sub(payload)
+
+        if not _auth_utils.is_emulator_enabled():
+            self._verify_kid_and_alg(header, payload)
```
Comment (hiranya911) [inline firebase_admin/_token_gen.py:396]: This changes the order of certain checks (aud and iss verified before header). Not a problem as far as the correctness of the functionality is concerned. But some of our tests might be unhappy with it depending on how they are implemented.

File: firebase_admin/_token_gen.py [hunk 3]
```diff
@@ -254,98 +275,134 @@ def __init__(self, **kwargs):
...
-            return verified_claims
+            return google.oauth2.id_token.verify_token(
+                token, request=request, audience=self.project_id, certs_url=self.cert_url)
         except google.auth.exceptions.TransportError as error:
             raise CertificateFetchError(str(error), cause=error)
         except ValueError as error:
-            if 'Token expired' in str(error):
-                raise self._expired_token_error(str(error), cause=error)
-            raise self._invalid_token_error(str(error), cause=error)
+            error_factory = (self._expired_token_error if 'Token expired' in str(error) else
```
Comment (hiranya911) [inline firebase_admin/_token_gen.py]: Nit: Can afford to leave this out of the diff.

File: firebase_admin/_auth_utils.py [hunk 4]
```diff
@@ -323,3 +340,72 @@ def _build_error_message(code, exc_type, custom_message):
...
+            **params: The optional additional parameters to substitute in the URL path.
+
+        Returns:
+            string: The corresponding resource URL.
+        """
+        return self._url_format.format(
+            api=api, project_id=self.project_id, version=self.version, **params)
+
+
+class TenantAwareAuthResourceUrlBuilder(AuthResourceUrlBuilder):
```
Comment (hiranya911) [inline firebase_admin/_auth_utils.py]: I didn't see any usages of this class. But seems useful if you ever decide to migrate the multitenancy code as well.

**Review by hiranya911 — COMMENTED** [review]
Looks pretty good. I called out a few things that can be optionally cleaned up.

PS: I would highly recommend adding some more unit tests to specifically cover the new functionality, and some sort of a CI flow to verify all tests with each commit.

File: firebase_admin/_token_gen.py (hunk 3 above)
Comment (brianrodri) [reply firebase_admin/_token_gen.py]: Done

File: firebase_admin/_token_gen.py (hunk 2 above)
Comment (brianrodri) [reply firebase_admin/_token_gen.py:396]: Ack'd, the tests as I ran them seemed to be OK so I don't think it's an issue.

File: firebase_admin/_auth_utils.py (hunk 4 above)
Comment (brianrodri) [reply firebase_admin/_auth_utils.py]: Yeah, I ended up reducing all of this cruft into a single function call, I've found it to be much nicer to work with.

File: firebase_admin/_auth_utils.py (hunk 1 above)
Comment (brianrodri) [reply firebase_admin/_auth_utils.py]: Yup, this is addressed with the new "way" I've decided to handle the URL building.

**Pull request closed without merging**